src/cargo/core/resolver/encode.rs

```diff
@@ -90,13 +92,56 @@ impl EncodableResolve {
                 try!(add_dependencies(id, pkg));
             }
         }
+        let mut metadata = self.metadata.unwrap_or(BTreeMap::new());
+
+        // Parse out all package checksums. After we do this we can be in a few
+        // situations:
+        //
+        // * We parsed no checksums. In this situation we're dealing with an old
+        //   lock file and we're gonna fill them all in.
+        // * We parsed some checksums, but not one for all packages listed. It
+        //   could have been the case that some were listed, then an older Cargo
+        //   client added more dependencies, and now we're going to fill in the
+        //   missing ones.
+        // * There are too many checksums listed, indicative of an older Cargo
+        //   client removing a package but not updating the checksums listed.
+        //
+        // In all of these situations they're part of normal usage, so we don't
+        // really worry about it. We just try to slurp up as many checksums as
+        // possible.
+        let mut checksums = HashMap::new();
+        let prefix = "checksum ";
+        let mut to_remove = Vec::new();
+        for (k, v) in metadata.iter().filter(|p| p.0.starts_with(prefix)) {
+            to_remove.push(k.to_string());
+            let k = &k[prefix.len()..];
+            let id: EncodablePackageId = try!(k.parse().chain_error(|| {
+                internal("invalid encoding of checksum in lockfile")
+            }));
+            let id = try!(to_package_id(&id.name,
+                                        &id.version,
+                                        id.source.as_ref(),
+                                        default,
+                                        &path_deps));
+            let v = if v == "<none>" {
+                None
+            } else {
+                Some(v.to_string())
+            };
+            checksums.insert(id, v);
+        }
+
+        for k in to_remove {
+            metadata.remove(&k);
+        }
```

Comment:
The checksum metadata is stored in a separate side table (`checksums`) and otherwise the `metadata` here is just preserved for future Cargos.

Right now nothing in Cargo reads `metadata`, it's just opaquely driven forward, so it shouldn't be confused at all.